### ISSUE TITLE
[Feat] improve performance significantly by selectively updating the Polygon options

### ISSUE THREAD

**Issue by nimlevy** [visgl/react-google-maps#524]
### Target Use Case

We are using the library to give our clients the ability to select zip code areas on the Google Map. Imagine a map marked with zip code boundaries, the user can hover to see details and then click to select the zip code (changing opacity and adding to a list)

### Proposal

The issue (somewhere between a bug and a feature) is that the dependencies of the [Polygon example](https://github.com/visgl/react-google-maps/blob/main/examples/geometry/src/components/polygon.tsx) (and probably the other geometry components as well) where the `polygon.setOptions` is called, are setting the options every time the mouse hovers over the polygon's area.
Updating the dependencies to the specific properties that would affect the rendering of the polygon (e.g. the `strokeOpacity`, `fillOpacity`, etc.) drastically improves performance.

I am not sure what is the best way to go about it, but the assumption stated in the following comment is, unfortunately, incorrect.
```
// update PolygonOptions (note the dependencies aren't properly checked
// here, we just assume that setOptions is smart enough to not waste a
// lot of time updating values that didn't change)
```
In other words - the `setOptions` is **not** smart enough.

Note: I have seen a drastic decrease in performance (color changes, info window's movement, etc. were lagging badly) when using the original format of paths (array of lat & lng) that React's dependency check could not properly compare

**Comment by usefulthink** [comment]
Like I said in #379, any help is very much appreciated! 

Especially since we will eventually move the geometry components (Polygon, Polyline and Circle) into the main library.

**Comment by nimlevy** [comment]
@usefulthink I'm not sure why, but I am not seeing the comment made by JalilSeif
But if they have any questions, or can use my code snippets, I can help.

**Comment by usefulthink** [comment]
Yeah, that must've been deleted in the meantime. It's a bit confusing that GitHub doesn't show that there was a deleted comment...

**Comment by samithahansaka** [comment]
I have this working locally, but there is a contract question I would rather settle with you before opening a PR.

First, a correction to the issue as written. The part about `setOptions` running on every hover is already fixed: the options object is now wrapped in `useMemoized(..., isDeepEqual)`, so an unchanged render never reaches `setOptions`. What is still there is that when one value does change, the whole options object is sent. Changing `fillOpacity` on hover also re-sends `editable`, `draggable`, `clickable` and every stroke and fill value, and writing `editable` makes the API rebuild the vertex and midpoint handles. That is the part that is expensive on shapes with many vertices.

The obvious fix is to send only the keys that changed. It works, and it is what I have built. But it gives up something the current code does by accident: today, any options change re-asserts every value, so if something else writes to the instance, the next update quietly snaps it back. With a diff, it never does.

That is not hypothetical. `examples/drawing/src/use-drawing-manager.tsx` calls `setOptions({editable: false, draggable: false, clickable: true})` directly on Polygon and Polyline instances, and the `polygon` and `polyline` props exist so an app can hold and mutate its own instance. The docs for those props currently say "all other props (paths, options, event handlers) will still be applied to this instance", which would stop being reliably true.

So there are two shapes this could take:

1. Diff everything. Biggest win, but the component stops reconciling an instance that anything else has written to.
2. Diff only the expensive keys (`editable`, `draggable`, `clickable`) and keep re-sending the rest. Still removes the handle rebuild, and everything else keeps healing itself as it does now.

I lean towards 2 for the geometry components, since it keeps the current contract intact, but I do not have numbers to say how much the cheap keys actually cost. I could not benchmark this properly, and the measurements in the original report predate the memoization fix, so they no longer describe the current code.

Which would you prefer? Either way the change also covers Polyline, Circle and Rectangle, and I have it applying options once on mount instead of twice, since the instances are already constructed with the full set.

One related thing while I was in there: `marker.tsx` and `use-map-options.ts` still carry the "we just assume that setOptions is smart enough to not waste a lot of time updating values that didn't change" comment. Happy to follow up there separately once this is settled.

**Comment by nimlevy** [comment]
Thank you for looking into this.
I ended up separating the `setOptions` and `setPaths` into two different `useEffect` hooks and that solved the performance issue; see code snippet at the end.

We won't the resources, at least in the near future, to test this.
But I would agree that your proposed 2nd option seem like a good choice.
Any downside of implementing the change regardless of my request?

```ts
// update PolygonOptions (note the dependencies aren't properly checked
// here, we just assume that setOptions is smart enough to not waste a
// lot of time updating values that didn't change)
useEffect(() => {
    polygon.setOptions(polygonOptions);
    // eslint-disable-next-line react-hooks/exhaustive-deps
}, [polygon, polygonOptions.fillOpacity, polygonOptions.strokeOpacity]);

useEffect(() => {
    if (!polygon || !polygonOptions.paths) {
        return
    }
    polygon.setPaths(polygonOptions.paths)
}, [polygon, polygonOptions.paths]);
```

**Comment by usefulthink** [comment]
@samithahansaka Thanks for the insight, I wasn't aware that those options are so expensive, even if the value doesn't actually change. It's probably also something to report to the Google Maps team.  

Generally, I'd say that code modifying the internal state of API objects that it doesn't own is asking for trouble – it has to expect that react-google-maps will update the state of the objects as it sees fit. Our internal design doesn't need to consider those cases at all if it helps the performance for the regular use-case. I have to look into the drawing-manager example and see why it's doing that, and if we can solve that in a different way (we probably can).

**Comment by samithahansaka** [comment]
@nimlevy on the downside question: there is one, and it is the thing @usefulthink just answered.

Sending only the changed keys means the component stops re-asserting the values it did not change. Today, any options change re-applies the whole object, so if something else has written to the shape, the next update quietly snaps it back. With a diff, it never does. That was my only reason for preferring the narrower option 2.

Given @usefulthink's point that the internal design does not need to accommodate code mutating objects it does not own, that objection goes away, so I would now go with option 1 and diff everything. It is the bigger win and it keeps the implementation simpler than special-casing which keys count as expensive.

Beyond that I do not see a downside. Passing a subset to `setOptions` is equivalent to passing the whole object, since the API leaves unspecified options untouched.

Two things worth noting from your snippet. Listing `polygonOptions.fillOpacity` and `polygonOptions.strokeOpacity` explicitly works, but it silently stops updating any other option, so `strokeWeight` or `zIndex` changes would no longer reach the shape. A diff gets you the same benefit without that tradeoff. And splitting `setPaths` out is a good catch: the library already does that, so `paths` is not part of the options object any more and is synced separately.

@usefulthink on reporting it upstream: worth doing. From reading the behaviour, the cost looks specific to `editable` and `draggable`, which rebuild the vertex and midpoint handles, rather than to `setOptions` generally. I have not profiled it, so that is inference from the API's behaviour rather than a measurement, and the original numbers in this issue predate the `useMemoized` fix so they no longer describe the current code.

I have the change ready against current `main`, covering Polygon, Polyline, Circle and Rectangle, and it also applies options once on mount instead of twice. I will open the PR shortly.

**Comment by nimlevy** [comment]
Great. Thank you!

**Comment by usefulthink** [comment]
@samithahansaka Awesome, thanks a lot for the PR, looking at it now. Do you, by any chance, have an example showing the time it takes to update the polygon options that can be used for a bug-report?

**Comment by samithahansaka** [comment]
@usefulthink no, I don't have one, and I should be clear that the `editable` and `draggable` cost is inferred rather than measured. I reasoned it from what those options have to do (rebuild the vertex and midpoint handles) and from the original report in this issue, but I never profiled it. The numbers @nimlevy posted in 2024 predate the `useMemoized` fix, so they no longer describe the current code either.

So the honest position is that the PR is safe on its own terms (sending a subset is equivalent, and it strictly reduces work handed to the API), but the specific claim that `editable` is the expensive part is not something I can currently back with a measurement.

Happy to build a repro for the bug report. What I would put together:

- a polygon with a few thousand vertices, `editable` on
- toggle only `fillOpacity` on an interval
- `performance.mark` around the `setOptions` call, comparing three cases: the full options object with `editable: true` included, the same minus `editable`, and just `{fillOpacity}`

If the difference is in the handle rebuild, the first case should be markedly slower than the third, and the second should tell us whether `draggable` and `clickable` behave the same way. If the three come out the same, then my explanation is wrong and the PR is a smaller win than described, which is worth knowing before it merges.

Would a standalone page in this repo work for that, or would you rather have it somewhere outside the project? I'd want to run it against a real API key rather than mocks, since this is exactly the kind of thing jsdom will not tell us anything useful about.

I'll get to your review comments on #1061 shortly. The `WeakMap` and accumulated-state approach is clearly better than what I wrote, particularly for correctly unsetting options that disappear.

**Comment by usefulthink** [comment]
> Happy to build a repro for the bug report. [...]

I think the main thing we'd have to show is that repeatedly setting the same option values isn't the zero-cost operation it should be. I'd also be interested in looking into what actually happens when options are re-applied.

> Would a standalone page in this repo work for that, or would you rather have it somewhere outside the project? I'd want to run it against a real API key rather than mocks, since this is exactly the kind of thing jsdom will not tell us anything useful about.

I typically use codesandbox.io for a repro like this (you can generate a demo-key here: https://mapsplatform.google.com/intl/de_de/maps-demo-key/), something like this: https://codesandbox.io/p/sandbox/jovial-sound-8jmrsh

**Comment by samithahansaka** [comment]
I ran it. Your framing was the right one to test, and the answer is yes: re-applying the same values is not free. But my explanation for why was wrong, so I want to correct that before it goes into a bug report.

Setup: one polygon, 3000 vertices, `editable: true`, on a real map. 40 iterations per case, medians below, `performance.now()` around each `setOptions` call.

### Re-applying unchanged values costs the same as changing them

| call | median |
| --- | --- |
| one **changed** value | 8.5 ms |
| one **unchanged** value | 8.5 ms |
| all 10 options, all unchanged | 48.4 ms |

There is no short-circuit. Writing a value costs the same whether or not it differs from what is already there, and the cost scales with how many keys are in the object:

| keys written (all unchanged) | median |
| --- | --- |
| 1 | 8.5 ms |
| 2 | 16.8 ms |
| 3 | 26.0 ms |
| 5 | 47.7 ms |

### It is not `editable`, which is what I assumed

I said the cost was `editable` and `draggable` rebuilding the vertex handles. That is not what happens.

| call | median |
| --- | --- |
| `editable: true` re-set on its own | **0.0 ms** |
| 5 stroke/fill options, unchanged | 43.6 ms |
| 6 flag options including `editable`, unchanged | 4.2 ms |

The handle count was 6001 before and after, so nothing is rebuilt. Re-setting `editable` is free.

The cost is per **stroke or fill** option: each one triggers a full redraw of the geometry, about 8.7 ms at 3000 vertices. The flags (`geodesic`, `visible`, `zIndex`, `clickable`, `draggable`, `editable`) are around 0.7 ms each.

That also explains something that looked odd: dropping `editable`, `draggable` and `clickable` from the object made it slightly *slower*, not faster. They were never the cost.

### What that means for the PR

It is a better change than I argued for, but for a different reason. The saving is avoiding redundant redraws from stroke and fill properties, so it applies to every geometry component whether or not the shape is editable, not just to editable ones.

I will fix the wrong explanation in the PR description.

### For the bug report

The clean statement is probably: `setOptions` applies every key it is given without comparing against the current value, and each stroke or fill key costs a full geometry redraw. So passing an unchanged option bag with N visual properties costs N redraws instead of none.

Happy to share the benchmark page, it is a single HTML file with the Maps script tag and no build step, if that is easier than a sandbox.

One caveat on the numbers: they came from an automated Chrome session on one machine, so treat the absolute milliseconds as indicative. The ratios were stable across three runs and the 0.0 ms for `editable` was not.

**Comment by usefulthink** [comment]
I had gemini build me a test-suite for this, testing every option including paths etc: https://codesandbox.io/p/sandbox/cocky-mountain-9gfjp5?file=%2Findex.html and to me this looks entirely like expected behavior. I can't see anything suspicious with the styling properties. At least on my machine, everything seems to be applied instantly.
